AutoinstallError exception

Chris-Peterson444 opened the pull request:
Adds a special exception for Autoinstall related failures and a
specific implementation for autoinstall validation failures.

When a user passes incorrect autoinstall data, the generated
bug report will be titled:

    "Autoinstall crashed with AutoinstallValidationError"

Where the traceback will start with a hint on the offending
section. For example:

    "Malformed autoinstall in 'apt' section"

Information from the original jsonschema ValidationError is
also available in the traceback to point to specific issues
with the provided data.

The section reporting is a little crude in the validation
of the base schema done by SubiquityServer, which can't discern
between the 'interative-sessions' and 'version' keys, but for now
the scope is pretty limited and can be fixed up at a later time.

Chris-Peterson444 commented:
Commit f6a7819ef9932ff45edbe6338f38137ff51fbb31 is a little big and I'm tempted to create a separate PR for it. Although it's mostly repeated lines to just test the the json schemas defined by the controllers are valid. 

Testing this is generally important, but `jsonschema.validate` can throw an error from either non-conforming data or bad json schemas. We should catch the latter at test time so only validation errors happen at run time.

Chris-Peterson444 commented:
I updated the commit adding `interactive-sections` in the base schema to also include it in the documentation's schema

Chris-Peterson444 commented:
Rebased to not conflict with https://github.com/canonical/subiquity/pull/1898

dbungert commented on subiquity/server/autoinstall.py:
```diff
@@ -0,0 +1,50 @@
+log = logging.getLogger("subiquity.server.autoinstall")
+
+
+class AutoinstallError(Exception):
```
Do we need to build out this functionality?  The result is that AutoinstallError doesn't really behave like other exceptions - str() and repr() look very different, we can get to the message on an exception with str(exception), etc.  Basically, what are the advantage of this approach over an declaration like:
```
class AutoinstallError(Exception):
    pass
```

Chris-Peterson444 replied:
You're right. I don't think any of the logic here in `AutoinstallError` is necessary. Mostly remnant of old ideas. Any special handling should be done by the subclass (like `__str__` in `AutoinstallValidationError).

dbungert commented on subiquity/server/controller.py:
```diff
@@ -54,6 +57,26 @@ async def _confirmed(self):
+                original_exception,
+            )
+
+            self.app.make_apport_report(
```
By default, autoinstall errors should not be apport reports.  This sort of flaw should move to manual reporting - we have this problem today were invalid data can trigger the bug flow and we don't want that.  Most invalid autoinstall is a user configuration error.  Select "invalid" autoinstall is really a bug, so we want to rely on manual user reporting for that case.

Chris-Peterson444 replied:
This is true. I'll flip this around and add some special handling in the `SubiquityServer` exception handler to not write out apport reports on instances of `AutoinstallError`.

Chris-Peterson444 commented on subiquity/server/autoinstall.py:
```diff
@@ -0,0 +1,50 @@
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.message!r}>"
+
+    def __str__(self):
```
Agreed, as above.

Chris-Peterson444 commented:
I simplified the new exception types a bit and added handling to explicitly not generate an apport report when failing on an instance of `AutoinstallError`.

dbungert commented on subiquity/server/server.py:
```diff
@@ -396,7 +404,8 @@ def make_apport_report(self, kind, thing, *, wait=False, **kw):
 
     async def _run_error_cmds(self, report):
-        await report._info_task
+        if hasattr(report, "_info_task"):
```
I'd like you to double-check this line.
`ErrorReport` unconditionally creates the _info_task attribute, so it should either be None or an asyncio task.   So it's always there, but may be None, and `await None` is bad.  Should this be `if report._info_task is not None` instead?

Chris-Peterson444 replied:
I didn't see the default value for the attribute, so your recommendation is right if `report` is of type `ErrorReport`. Although in the case of `AutoinstallError` exceptions I'm passing `report=None`. I was trying to catch both cases here, but I should have been more explicit.

My assumption is that the `ErrorReport` class is about generating apport reports based on the exception, but here we don't want to do that.

What do you think of rewriting as:
```python
async def _run_error_cmds(self, report: Optional[ErrorReport] = None):
    if report is not None and report._info_task is not None:
        await report._info_task
     ...
```

dbungert replied:
Ah, report=None would also have been hosed (`getattr(None, ...)` is also bad.)
I expect that ErrorReport is the correct type here, so your latest suggestion looks correct.

dbungert reviewed (commented):
One item to review then I think this is ready.
